File: Homework/HW12/HW12.py
from typing import List


class HW12:
    def __init__(self):
        self.parent = []

    def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
        self.parent = list(range(n + 1))

        pq = []
        for i in range(n):
            pq.append((wells[i], 0, i + 1))

        for pipe in pipes:
            pq.append((pipe[2], pipe[0], pipe[1]))

        pq.sort()

        total_cost = 0
        for cost, house1, house2 in pq:
            if self.union(house1, house2):
                total_cost += cost

        return total_cost

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX != rootY:
            self.parent[rootY] = rootX
            return True
        return False

File: Homework/HW12/HW12.java
package Homework.HW12;

import java.util.PriorityQueue;
import java.util.Comparator;

public class HW12 {
    private int[] parent;

    public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        parent = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            parent[i] = i;
        }

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));

        for (int i = 0; i < n; i++) {
            pq.offer(new int[] { wells[i], 0, i + 1 });
        }

        for (int[] pipe : pipes) {
            pq.offer(new int[] { pipe[2], pipe[0], pipe[1] });
        }

        int totalCost = 0;
        while (!pq.isEmpty()) {
            int[] edge = pq.poll();
            if (union(edge[1], edge[2])) {
                totalCost += edge[0];
            }
        }

        return totalCost;
    }

    private int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    private boolean union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX != rootY) {
            parent[rootY] = rootX;
            return true;
        }
        return false;
    }
}
